### ISSUE TITLE
Support stripping away layers in `iterate_many`

### ISSUE THREAD

**Issue by dominiklohmann** [simdjson/simdjson#2252]
I have a potentially infinite array of objects I want to parse:

```
[{…}, {…}, {…}, …]
```

I do not really care about the outer array, but am much rather looking for the _fastest_ way to stream the objects contained in the array.

The popular command-line tool `jq` supports a way to "strip" layers of JSON while streaming the output without requiring the outer array's structure to be fully available, i.e., it continues to work even if the end of the array never arrives.

```bash
jq -ercn --stream 'fromstream(1|truncate_stream(inputs))'
```

I am looking to do something just like this in my own application, which uses simdjson's `iterate_many` API. My (perhaps naïve) thinking is that the API likely needs an additional parameter that specifies how many layers to strip away.

**Comment by lemire** [comment]
We already support comma-separated documents. So all that is needed is to strip the starting `[` and the ending `]` which is nearly trivial.

Hmmm. So yeah, this could be implemented, very likely, with maybe 10 to 50 lines of code and a few hours of hacking mostly spent on documentation and testing.

**Comment by dominiklohmann** [comment]
For reference, here's the PR that added the support for comma separated documents:
- #2016

How would you recommend integrating this into `iterate_many`? I don't think it's particularly clean to add yet another argument, and changing the boolean added in #2016 to an enum is a breaking change, unless I add another set of overloads to the API.

**Comment by lemire** [comment]
> How would you recommend integrating this into iterate_many?

I am not sure. I am opening this for discussion.

Note that simdjson is a community-based project.

**Comment by prashu0705** [comment]
Hi @lemire  and @dominiklohmann ,

I’d love to contribute to this issue!

The task seems to involve enabling `iterate_many` to process input where we strip the outer array brackets `[` and `]`, allowing for streaming of the objects within. This would help in scenarios where the input stream could be infinite or partial, and we don't need to process the outer structure.

To solve this, I propose introducing a new configuration flag (e.g., `strip_array`) or using an optional preprocessing step to remove the brackets before proceeding with the existing iteration logic. This would integrate seamlessly with the current functionality and maintain efficiency.

This issue would be an excellent opportunity to contribute to simdjson.

Could you please assign this issue to me? I’m excited to get started!

Thanks in advance!

**Comment by lemire** [comment]
@prashu0705 Sure. Note that we want to preserve backward compatibility. Thus I would prefer if you we do not change the existing functions. (It is bad form to break existing code.)

**Comment by shikharish** [comment]
Hello @lemire , what is the status of this issue? I would like to work on it.

**Comment by lemire** [comment]
@shikharish To my knowledge, nobody has been working on this. I do not think it is difficult.

**Comment by shikharish** [comment]
@lemire I don't think this will be that trivial. comma-separted `iterate_many` currently has a constrain on the batch_size:
> We also support comma-separated documents, but with some performance limitations. The `iterate_many` function  takes in an option to allow parsing of comma separated documents (which defaults on false). In this mode, the entire buffer is processed in one batch. Therefore, the total size of the document should not exceed the maximal capacity of the parser (4 GB). This mode also effectively disallow multithreading. It is therefore mostly suitable for not "very large" inputs. In this mode, the batch_size parameter is effectively ignored, as it is set to at least the document size.

How will this handle large arrays then?

**Comment by lemire** [comment]
@shikharish If you want to support arrays that are beyond 4 GB, then yes, additional work is needed. That part is not entirely trivial, but it should not be too difficult engineering wise.

**Comment by shikharish** [comment]
Even for arrays less than 4GB, isn't it very inefficient? Because it will allocate memory for the whole array size, so it isn't exactly "streaming"?

**Comment by lemire** [comment]
@shikharish The standard way to implement streaming JSON is by Newline-delimited JSON. We received a contribution that added support for comma-separated inputs, with some limitations. Whether is good enough or not, I do not know. It seems that it was good enough for the team that contributed it. One should look at the actual applications.

The issue is an algorithmic matter. If I give you a block of data within a stream of several JSON documents, separated by spaces or line feeds, and with a block that may start in the middle of a JSON document or end in the middle of another one... We have an algorithm that can reliably identify the locations of the documents. It works well.

If you have comma separated documents, you need another such algorithm, which nobody has developed yet (at least not in simdjson).

In all cases, someone needs to do the work, design the efficient algorithms, tests them, benchmark them and so forth. These issues only get resolved by writing code.

**Comment by lemire** [comment]
Thanks to @jaja360's work, we will soon have a fix for this issue, see https://github.com/simdjson/simdjson/pull/2674

**Comment by lemire** [comment]
It will be part of the next release.
